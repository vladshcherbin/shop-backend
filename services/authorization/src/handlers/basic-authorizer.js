import logger from '../modules/logger'
import generatePolicy from '../modules/policy'

export async function handler(event, context, callback) {
  try {
    logger.info({ event }, 'start basic authorizer')

    const { authorizationToken, methodArn, type } = event

    if (type !== 'TOKEN') {
      return callback('Unauthorized')
    }

    const [authType, token] = authorizationToken.split(' ')

    if (authType !== 'Basic' || !token) {
      return callback('Unauthorized')
    }

    const [user, password] = Buffer.from(token, 'base64').toString('utf8').split(':')

    if (!user || !password) {
      return callback('Unauthorized')
    }

    return (process.env[user] === password)
      ? callback(null, generatePolicy(token, 'Allow', methodArn))
      : callback(null, generatePolicy(token, 'Deny', methodArn))
  } catch (error) {
    logger.error({ error }, 'authorizer error')

    return callback('Unauthorized')
  } finally {
    logger.info('finish basic authorizer')
  }
}
