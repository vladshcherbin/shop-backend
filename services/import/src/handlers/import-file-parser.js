import { GetObjectCommand, S3Client } from '@aws-sdk/client-s3'
import logFile from '../modules/csv'
import logger from '../modules/logger'
import { errorResponse, successfulResponse } from '../modules/response'

export async function handler(event) {
  try {
    logger.info({ event }, 'file created')

    const [record] = event.Records
    const client = new S3Client({ region: record.awsRegion })
    const { Body } = await client.send(new GetObjectCommand({
      Bucket: record.s3.bucket.name,
      Key: record.s3.object.key
    }))

    await logFile(Body)

    return successfulResponse({ status: 'success' })
  } catch (error) {
    return errorResponse(error)
  }
}
