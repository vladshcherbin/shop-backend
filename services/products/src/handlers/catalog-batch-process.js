import { PublishCommand, SNSClient } from '@aws-sdk/client-sns'
import knexConnection from 'knex'
import logger from '../modules/logger'

const knex = knexConnection({
  client: 'pg',
  connection: {
    host: process.env.DB_HOST,
    port: process.env.DB_PORT,
    user: process.env.DB_USER,
    password: process.env.DB_PASSWORD,
    database: process.env.DB_DATABASE,
    ssl: {
      rejectUnauthorized: false
    }
  }
})

export async function handler(event) {
  try {
    logger.info({ event }, 'catalog batch process start')

    if (event.Records.length) {
      const products = event.Records.map((record) => JSON.parse(record.body))

      logger.info({ products }, 'saving products to database')

      await knex.transaction(async (transaction) => {
        const currentDate = new Date().toISOString()
        const insertedProducts = await transaction('products')
          .insert(products.map((product) => ({
            title: product.title,
            description: product.description,
            price: product.price,
            image_url: product.image_url,
            created_at: currentDate,
            updated_at: currentDate
          })))
          .returning('*')

        logger.info({ insertedProducts }, 'inserted products')

        const productsWithCount = products.filter((product) => product.count)

        if (productsWithCount.length) {
          const insertedStocks = await transaction('stocks')
            .insert(productsWithCount.map((product) => ({
              product_id: insertedProducts.find((insertedProduct) => (
                insertedProduct.title === product.title
              )).id,
              count: product.count,
              created_at: currentDate,
              updated_at: currentDate
            })))

          logger.info({ insertedStocks }, 'inserted stocks')
        }

        return insertedProducts
      })

      logger.info('sending notification email')

      const snsClient = new SNSClient({ region: 'eu-west-1' })

      await snsClient.send(new PublishCommand({
        TopicArn: process.env.SNS_ARN,
        Subject: 'New products added',
        Message: JSON.stringify(products),
        MessageAttributes: {
          count: {
            DataType: 'String',
            StringValue: products.length > 1 ? 'many' : 'single'
          }
        }
      }))
    }

    logger.info('catalog batch process end')
  } catch (error) {
    logger.error({ error }, 'processing error')
  }
}
