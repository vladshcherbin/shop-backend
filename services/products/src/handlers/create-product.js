import knexConnection from 'knex'
import { number, object, string } from 'yup'
import logger from '../modules/logger'
import { errorResponse, successfulResponse } from '../modules/response'
import { validate } from '../modules/validation'

const knex = knexConnection({
  client: 'pg',
  connection: {
    host: process.env.DB_HOST,
    port: process.env.DB_PORT,
    user: process.env.DB_USER,
    password: process.env.DB_PASSWORD,
    database: process.env.DB_DATABASE,
    ssl: {
      rejectUnauthorized: false
    }
  }
})

export async function handler(event) {
  try {
    logger.info({ event }, 'create product')

    const data = JSON.parse(event.body)
    const schema = object({
      title: string().required().min(3),
      description: string(),
      price: number().positive(),
      image_url: string().required().url(),
      count: number().integer().positive()
    })

    const { count, ...restValidInput } = await validate(schema, data)
    const product = await knex.transaction(async (transaction) => {
      const currentDate = new Date().toISOString()
      const [insertedProduct] = await transaction('products')
        .insert({
          ...restValidInput,
          created_at: currentDate,
          updated_at: currentDate
        })
        .returning('*')

      await transaction('stocks').insert({
        product_id: insertedProduct.id,
        count,
        created_at: currentDate,
        updated_at: currentDate
      })

      return insertedProduct
    })

    logger.info({ product }, 'product created')

    return successfulResponse({ product })
  } catch (error) {
    return errorResponse(error)
  }
}
