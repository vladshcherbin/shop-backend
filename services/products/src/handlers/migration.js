import knexConnection from 'knex'
import logger from '../modules/logger'

const knex = knexConnection({
  client: 'pg',
  connection: {
    host: process.env.DB_HOST,
    port: process.env.DB_PORT,
    user: process.env.DB_USER,
    password: process.env.DB_PASSWORD,
    database: process.env.DB_DATABASE,
    ssl: {
      rejectUnauthorized: false
    }
  }
})

export async function handler() {
  try {
    await knex.migrate.latest()
  } catch (error) {
    logger.error({ error }, 'Product migration failed')
  }
}
