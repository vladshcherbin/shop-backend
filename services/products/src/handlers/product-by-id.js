import { findById } from '../data/products'

export async function handler(event) {
  const product = await findById(Number(event?.pathParameters?.id))

  if (!product) {
    return {
      statusCode: 404,
      headers: {
        'Access-Control-Allow-Origin': '*'
      },
      body: JSON.stringify({ message: 'Product not found' })
    }
  }

  return {
    statusCode: 200,
    headers: {
      'Access-Control-Allow-Origin': '*'
    },
    body: JSON.stringify({ product })
  }
}
