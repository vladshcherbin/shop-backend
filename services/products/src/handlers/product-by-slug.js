import products from '../data/products'

export async function handler(event) {
  const { slug } = event.pathParameters

  return {
    statusCode: 200,
    body: JSON.stringify({
      product: products.find((product) => product.slug === slug)
    })
  }
}
