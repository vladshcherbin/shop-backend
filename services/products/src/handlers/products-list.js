import knexConnection from 'knex'
import logger from '../modules/logger'
import errorResponse from '../modules/response/error-response'
import successfulResponse from '../modules/response/successful-response'

const knex = knexConnection({
  client: 'pg',
  connection: {
    host: process.env.DB_HOST,
    port: process.env.DB_PORT,
    user: process.env.DB_USER,
    password: process.env.DB_PASSWORD,
    database: process.env.DB_DATABASE,
    ssl: {
      rejectUnauthorized: false
    }
  }
})

export async function handler(event) {
  try {
    logger.info({ event }, 'get products')

    const products = await knex
      .select('products.id', 'products.title', 'products.description', 'products.price', 'products.image_url', 'stocks.count')
      .from('products')
      .innerJoin('stocks', 'products.id', 'stocks.product_id')
      .orderBy('products.created_at', 'desc')

    return successfulResponse({ products })
  } catch (error) {
    return errorResponse(error)
  }
}
