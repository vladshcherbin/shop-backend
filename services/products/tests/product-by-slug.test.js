import { handler } from '../src/handlers/product-by-slug'

test('Correct product is returned', async () => {
  const response = await handler({
    pathParameters: { slug: 'basic-top' }
  })

  expect(response.statusCode).toBe(200)
  expect(response.body).toMatchSnapshot()
})

test('Not found error if slug is missing', async () => {
  const response = await handler()

  expect(response.statusCode).toBe(404)
  expect(response.body).toMatchSnapshot()
})

test('Not found error if slug is empty', async () => {
  const response = await handler({
    pathParameters: { slug: '' }
  })

  expect(response.statusCode).toBe(404)
  expect(response.body).toMatchSnapshot()
})

test('Not found error if product with slug doesn\'t exist', async () => {
  const response = await handler({
    pathParameters: { slug: 'missing' }
  })

  expect(response.statusCode).toBe(404)
  expect(response.body).toMatchSnapshot()
})
