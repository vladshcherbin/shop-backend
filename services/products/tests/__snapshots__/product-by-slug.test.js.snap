// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Correct product is returned 1`] = `"{\\"product\\":{\\"id\\":2,\\"title\\":\\"Basic top\\",\\"description\\":\\"round neck, wide straps\\",\\"image\\":\\"https://www.sinsay.com/media/catalog/product/cache/850/a4e40ebdc3e371adff845072e1c73f37/2/6/2628F-18X-001-1_3.jpg\\",\\"slug\\":\\"basic-top\\",\\"price\\":9.99,\\"count\\":6}}"`;

exports[`Not found error if product with slug doesn't exist 1`] = `"{\\"message\\":\\"Product not found\\"}"`;

exports[`Not found error if slug is empty 1`] = `"{\\"message\\":\\"Product not found\\"}"`;

exports[`Not found error if slug is missing 1`] = `"{\\"message\\":\\"Product not found\\"}"`;
